sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AnsiTypeCoercion.scala

```diff
@@ -0,0 +1,316 @@
+ *   * Comparison is allowed iff the data type precedence list of both sides has at least one common
+ *     element. When evaluating the comparison, Spark casts both sides as the tightest common data
+ *     type of their precedent lists.
+ *   * There should be at least one common data type among all the children's precedence lists for
+ *     the following operators. The data type of the operator is the tightest common precedent
+ *     data type.
+ *       * In
+ *       * Except(odd)
+ *       * Intersect
+ *       * Greatest
+ *       * Least
+ *       * Union
+ *       * If
+ *       * CaseWhen
+ *       * CreateArray
+ *       * Array Concat
+ *       * Sequence
+ *       * MapConcat
+ *       * CreateMap
+ *   * For complex types (struct, array, map), Spark recursively looks into the element type and
+ *     applies the rules above. If the element nullability is converted from true to false, add
+ *     runtime null check to the elements.
+ *  Note: this new type coercion system will allow implicit converting String type literals as other
+ *  primitive types, in case of breaking too many existing Spark SQL queries. This is a special
+ *  rule and it is not from the ANSI SQL standard.
+ */
+object AnsiTypeCoercion extends TypeCoercionBase {
+  override def typeCoercionRules: List[Rule[LogicalPlan]] =
+    InConversion ::
+      WidenSetOperationTypes ::
+      PromoteStringLiterals ::
+      DecimalPrecision ::
+      FunctionArgumentConversion ::
+      ConcatCoercion ::
+      MapZipWithCoercion ::
+      EltCoercion ::
+      CaseWhenCoercion ::
+      IfCoercion ::
+      StackCoercion ::
+      Division ::
+      IntegralDivision ::
+      ImplicitTypeCasts ::
+      DateTimeOperations ::
+      WindowFrameCoercion ::
+      StringLiteralCoercion ::
+      Nil
+
+  /**
+   * Find the tightest common type of two types that might be used in a binary expression.
+   */
+  override def findTightestCommonType(t1: DataType, t2: DataType): Option[DataType] = {
+    (t1, t2) match {
+      case (t1, t2) if t1 == t2 => Some(t1)
+      case (NullType, t1) => Some(t1)
+      case (t1, NullType) => Some(t1)
+
+      case (t1: NumericType, t2: NumericType) =>
+        findTightestCommonNumericType(t1, t2)
+
+      case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
+        Some(TimestampType)
+
+      case (t1, t2) => findTypeForComplex(t1, t2, findTightestCommonType)
+    }
+  }
+
+  @tailrec
+  private def findTightestCommonNumericType(t1: NumericType, t2: NumericType): Option[DataType] = {
+    (t1, t2) match {
+      case (i: IntegralType, d: DecimalType) =>
+        if (d.isWiderThan(i)) {
+          Some(t2)
+        } else {
+          findTightestCommonNumericType(DecimalType.forType(i), d)
+        }
+
+      case (t1: DecimalType, t2: IntegralType) =>
+        findTightestCommonNumericType(t2, t1)
+
+      case (t1: DecimalType, t2: DecimalType) =>
+        Some(DecimalPrecision.widerDecimalType(t1, t2))
```

Comment:
What's a relationship between `AnsiTypeCoercion` and `DecimalPrecision`? In the existing `DecimalPrecision`, it seems all the `(decimal, decimal)` cases are handled in `DecimalPrecision` though.

Reply:
Here it I am merging `findWiderTypeForDecimal` into `findTightestCommonType`, to make the logic more clear.
For the rule `DecimalPrecision`, it decides the result data type of various operators involving decimals, e.g. `Add`, `Subtract`.

Reply:
Ah, ok. The latest change looks good.